pydantic/dataclasses.py

```diff
@@ -0,0 +1,43 @@
+from dataclasses import dataclass as dc
+from .main import validate_model, create_model
+
+
+def pydantic_model(cls):
+    fields = {name: (field.type, field.default)
+              for name, field
+              in cls.__dataclass_fields__.items()}
+    model = create_model('', **fields)
+    return model
+
+
+def dataclass(_cls=None, *, init=True, repr=True, eq=True,
+              order=False, unsafe_hash=False, frozen=False):
+    def wrap(cls):
+        post_init_orig = getattr(cls, '__post_init__', lambda self: None)
+        cls.__post_init__ = post_init_orig
+        cls = dc(init=init, repr=repr, eq=eq, order=order,
+                 unsafe_hash=unsafe_hash, frozen=frozen)(cls)
+        model = pydantic_model(cls)
+
+        def set_attr(self, name, value):
+            super(cls, self).__setattr__(name, value)
+            if getattr(self, '__initialized', False):
+                validate_model(model, self.__dict__)
```

Comment:
as with [a model](https://github.com/samuelcolvin/pydantic/blob/master/pydantic/main.py#L171) we should be able to validate one field without rerunning the whole of validation.

Reply:
so (with the comment below), something like

```py
_, error_ = self.__model__.fields[name].validate(value, self.__model__.dict(exclude={name}), loc=name)
if error_:
    raise ValidationError([error_])
```

Also perhaps we should move

```py
super(cls, self).__setattr__(name, value)
```

to after the validation?

Reply:
it could be done with the model in the closure as well. Maybe I can throw away the __post_init__ method and validate only the attributes even during init?

Reply:
yes, you could keep everything in closure not associated with the cls, except that would be annoying if anyone wanted to access the model from their own methods on the cls.

If if you're keeping everything in closure, that should include `__initialized` too I think.